### PR TITLE
[SPIR-V][Headers] Enable programs that include system headers on Windows for SPIRV32 and SPIRV64 targets

### PR DESCRIPTION
**Pull request by asudarsa**
This is text from OpenCL SPIR Spec 2.0 (https://registry.khronos.org/SPIR/specs/spir_spec-2.0.pdf)

The OpenCL size_t, ptrdiff_t,uintptr_t, intptr_t, atomic_size_t, atomic_ptrdiff_t, atomic_uintptr_t and atomic_intptr_t data types are mapped to LLVM i32 when the device address width is equal to 32 bits and to LLVM i64 when the device address width is equal 64 bits.

After some analysis though, I see that the windows os based customizations made for the SPIR target (https://github.com/intel/llvm/pull/325) are necessary for successful compilation on windows.
Also, they do not seem to be violating the spec. It's just that 'long' and 'long long' mean different things on windows and linux. On 64-bit linux, long type has 8 bytes, but on 64-bit windows, long type has 4 bytes.
Hence, I have extended the SPIR target changes to the SPIR-V target as well. 
This PR extends those changes for SPIRV32 and SPIRV64 targets and added appropriate tests as well.

We would eventually want to submit this (and PR #325) upstream. 

Thanks

### CODE REVIEW

**Review by sarnex — APPROVED** [review]
The relevant spec paragraph seems to be

```
The OpenCL size t,ptrdiff t,uintptr t, intptr t, atomic size t, atomic ptrdiff t,
atomic uintptr t and atomic intptr t data types are mapped to LLVM i32 when the
device address width is equal to 32 bits and to LLVM i64 when the device address width is
equal 64 bits
```

This change looks correct to me but I will also ask the CFE team to double check.

**Review by MrSidims — APPROVED** [review]

**Comment by sarnex** [comment]
We may want to submit this upstream.

**Comment by asudarsa** [comment]
> We may want to submit this upstream.
Thanks for reviews all.
This customization for windows is present only downstream. Also, the person who added it is no longer at Intel. Will wait for front end response.

thanks

**Comment by elizabethandrews** [comment]
@mikerice1969 can you take a look at this? I am not very familiar with this and @premanandrao is OOO this week

**Comment by asudarsa** [comment]
Thanks @elizabethandrews for your comment. Will wait for feedback.

**Comment by asudarsa** [comment]
Will mark it 'ready for review' once testing is complete. Thanks

File: clang/lib/Basic/Targets/SPIR.h
Comment (zahiraam) [inline clang/lib/Basic/Targets/SPIR.h:516]: This TBD is coming from copying the above code, right? Doesn't need to be addressed here.

**Review by zahiraam — APPROVED** [review]
LGTM. Thanks.

**Comment by premanandrao** [comment]
@asudarsa, the changes look okay to me, but I would like there to be tests for the sizes of the various types (particularly, those that may have changed) when using the different triples.   Also tests for the macro definitions.

**Comment by asudarsa** [comment]
> @asudarsa, the changes look okay to me, but I would like there to be tests for the sizes of the various types (particularly, those that may have changed) when using the different triples. Also tests for the macro definitions.

Thanks @premanandrao for the feedback. I will provide a resolution soon.

**Comment by asudarsa** [comment]
> > @asudarsa, the changes look okay to me, but I would like there to be tests for the sizes of the various types (particularly, those that may have changed) when using the different triples. Also tests for the macro definitions.
> 
> Thanks @premanandrao for the feedback. I will provide a resolution soon.

Added more testing in latest commit. Thanks

**Comment by asudarsa** [comment]
hi @sarnex and @MrSidims can you please take a look again? It seems SPIR/SPIR64 customizations in PR #325 are necessary for correct functioning on windows. I have added SPIRV32/SPIRV64 customizations now.

Thanks

**Review by sarnex — APPROVED** [review]
I can't review the original patch in detail but I approve of the idea if reapply it if it's necessary.

**Review by premanandrao — APPROVED** [review]
Thanks, lgtm.

**Pull request merged**
